search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,587 @@
+                      gpointer                   user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    gchar *joined_terms = g_strjoinv (" ", terms);
+
+    yelp_launch (joined_terms, app, invocation, "show-search");
+
+    g_free (joined_terms);
+
+    return TRUE;
+}
+
+static gboolean
+handle_activate_result (YelpShellSearchProvider2  *skeleton,
+                        GDBusMethodInvocation     *invocation,
+                        gchar                     *result,
+                        gchar                    **terms,
+                        guint32                    timestamp,
+                        gpointer                   user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+
+    yelp_launch (result, app, invocation, "show-page");
+
+    return TRUE;
+}
+
+static void
+search_provider_app_dispose (GObject *obj)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (obj);
+
+    g_clear_object (&self->skeleton);
+    g_clear_pointer (&self->page_data_hash_map, g_hash_table_unref);
+
+    G_OBJECT_CLASS (yelp_search_provider_app_parent_class)->dispose (obj);
+}
+
+static void
+preload_data_cb (YelpDocument          *document,
+                 YelpDocumentSignal     signal,
+                 YelpSearchProviderApp *self,
+                 GError                *error)
+{
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+    gint i;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_ERROR) {
+        g_warning ("error during preloading data: %s", (error != NULL) ? error->message : "unknown error");
+        return;
+    }
+
+    if (signal == YELP_DOCUMENT_SIGNAL_CONTENTS)
+        return;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_INFO) {
+        for (iter = page_ids; *iter; iter++) {
+            gchar *page_id = *iter;
+            PageData *data;
+            gchar *icon_string = yelp_document_get_page_icon (document, page_id);
+            GIcon *icon = g_themed_icon_new (icon_string);
+            g_free (icon_string);
+
+            data = page_data_new_steal (yelp_document_get_page_title (document, page_id), 
+                                        yelp_document_get_page_desc (document, page_id), 
+                                        icon);
+            g_hash_table_insert (self->page_data_hash_map, page_id, data);
+        }
+
+        g_application_release (G_APPLICATION (self));
+
+        for (i = 0; i < self->delayed_result_getters->len; i++) {
+            DelayedResultGetter *delayed = g_ptr_array_index (self->delayed_result_getters, i);
+
+            g_dbus_method_invocation_return_value (delayed->invocation,
+                                                   delayed->result_getter (delayed->terms,
+                                                                           self->page_data_hash_map));
+        }
+
+        g_clear_pointer (self->delayed_result_getters, g_ptr_array_unref);
```

Comment:
Also remember to pass an address to `g_clear_pointer` (so `&self->foo`, not just `self->foo`).